Framework/CurveFitting/src/Functions/CrystalFieldFunction.cpp

```diff
@@ -0,0 +1,1320 @@
+    return m_control.isExplicitlySet(i);
+  } else if (i < m_nControlSourceParams) {
+    return m_source->isExplicitlySet(i - m_nControlParams);
+  } else {
+    return m_target->isExplicitlySet(i - m_nControlSourceParams);
+  }
+}
+
+/// Get the fitting error for a parameter
+double CrystalFieldFunction::getError(size_t i) const {
+  checkSourceFunction();
+  checkTargetFunction();
+  if (i < m_nControlParams) {
+    return m_control.getError(i);
+  } else if (i < m_nControlSourceParams) {
+    return m_source->getError(i - m_nControlParams);
+  } else {
+    return m_target->getError(i - m_nControlSourceParams);
+  }
+}
+
+/// Set the fitting error for a parameter
+void CrystalFieldFunction::setError(size_t i, double err) {
+  checkSourceFunction();
+  checkTargetFunction();
+  if (i < m_nControlParams) {
+    m_control.setError(i, err);
+  } else if (i < m_nControlSourceParams) {
+    m_source->setError(i - m_nControlParams, err);
+  } else {
+    m_target->setError(i - m_nControlSourceParams, err);
+  }
+}
+
+/// Change status of parameter
+void CrystalFieldFunction::setParameterStatus(
+    size_t i, IFunction::ParameterStatus status) {
+  checkSourceFunction();
+  checkTargetFunction();
+  if (i < m_nControlParams) {
+    m_control.setParameterStatus(i, status);
+  } else if (i < m_nControlSourceParams) {
+    m_source->setParameterStatus(i - m_nControlParams, status);
+  } else {
+    m_target->setParameterStatus(i - m_nControlSourceParams, status);
+  }
+}
+
+/// Get status of parameter
+IFunction::ParameterStatus
+CrystalFieldFunction::getParameterStatus(size_t i) const {
+  checkSourceFunction();
+  checkTargetFunction();
+  if (i < m_nControlParams) {
+    return m_control.getParameterStatus(i);
+  } else if (i < m_nControlSourceParams) {
+    return m_source->getParameterStatus(i - m_nControlParams);
+  } else {
+    return m_target->getParameterStatus(i - m_nControlSourceParams);
+  }
+}
+
+/// Return parameter index from a parameter reference.
+size_t
+CrystalFieldFunction::getParameterIndex(const ParameterReference &ref) const {
+  checkSourceFunction();
+  checkTargetFunction();
+  if (ref.getLocalFunction() == this) {
+    return ref.getLocalIndex();
+  }
+  auto index = m_control.getParameterIndex(ref);
+  if (index < m_nControlParams) {
+    return index;
+  }
+  index = m_source->getParameterIndex(ref);
+  if (index < m_source->nParams()) {
+    return index + m_nControlParams;
+  }
+  return m_target->getParameterIndex(ref) + m_nControlSourceParams;
+}
+
+/// Set up the function for a fit.
+void CrystalFieldFunction::setUpForFit() {
+  checkSourceFunction();
+  updateTargetFunction();
+  IFunction::setUpForFit();
+}
+
+/// Declare a new parameter
+void CrystalFieldFunction::declareParameter(const std::string &, double,
+                                            const std::string &) {
+  throw Kernel::Exception::NotImplementedError(
+      "CrystalFieldFunction cannot not have its own parameters.");
```

Comment:
the double negative makes this error message a bit tricky to understand

Reply:
:)